### PR TITLE
Adding 2023 09 01 for nginx

### PR DESCRIPTION
**Pull request by zaowang-ms**
# ARM (Control Plane) API Specification Update Pull Request 

> Tip: overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

diagram

<details>
<summary> Click here to see the details of Step 1 </summary>

### Breaking changes review (Diagram Step 1)

If the automation determines you have breaking changes, i.e. Step 1 from the diagram applies to you,
  you must follow the [breaking changes process].  
  **IMPORTANT** This applies even if:
- The tool fails while it shouldn't, e.g. due to runtime exception, or incorrect detection of breaking changes.
- You believe there is no need for you to request breaking change approval, for any reason. 
    Such claims must be reviewed, and the process is the same.

</details>

<details> 
<summary> Click here to see the details of Step 2 </summary>

### ARM API changes review (Diagram Step 2)

- If this PR is in purview of ARM review then automation will add the `ARMReview` label.
- If you want to force ARM review, add the label yourself.
- Proceed according to the diagram at the top of this comment.

</details>

<details>
<summary> Click here to see the diagram footnotes</summary>

### Diagram footnotes

[1] [ARM review queue] (for **merge** queues, see [2])  
The PRs are processed by time opened, ascending. Your PR may show up on 2nd or later page. 
If you addressed Step 1 from the diagram and your PR is not showing up in the queue, ensure the label `ARMChangesRequested` 
is removed from your PR. This should cause the label `WaitForARMFeedback` to be added.
[2] [public repo merge queue], [private repo merge queue] (for **ARM review** queue, [1])

If you need further help with anything, see `Getting help` section below.

</details>

## Purpose of this PR

What's the purpose of this PR? Check all that apply. This is **mandatory**!

  - [x] New API version. (If API spec is not defined in TypeSpec, the PR should have been generated using [OpenAPI Hub](https://aka.ms/openapihub)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix swagger quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the Diagram Step 2, "ARM API changes Review", for this PR.

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://aka.ms/azsdk/pr-suppressions) 
to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

[ARM review queue]: https://github.com/search?q=org%3AAzure+is%3Apr+is%3Aopen+label%3AWaitForARMFeedback+-label%3AIDCDevDiv++draft%3Afalse+sort%3Acreated-asc+&type=pullrequests
[public repo merge queue]: https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+draft%3Afalse+sort%3Acreated-asc
[private repo merge queue]: https://github.com/Azure/azure-rest-api-specs-pr/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+-label%3AApproved-OkToMerge+draft%3Afalse+sort%3Acreated-asc
[breaking changes process]: https://eng.ms/docs/cloud-ai-platform/azure-core/azure-core-pm-and-design/trusted-platform-pm-karimb/service-lifecycle-and-actions-team/service-lifecycle-actions-team/apex/media/launchingproductbreakingchanges#breaking-change-process-1

### CODE REVIEW

**Comment by zaowang-ms** [comment]
Sync version from private to public. Code updates are exactly same as it is in private repo here: https://github.com/Azure/azure-rest-api-specs-pr/tree/RPSaaSMaster/specification/nginx/resource-manager/NGINX.NGINXPLUS/stable/2023-09-01

**Comment by zaowang-ms** [comment]
Fixed commit history to make 2023-04-01 as first commit, and 2023-09-01 as follows

**Comment by mikekistler** [comment]
Breaking changes previously reviewed and approved in https://github.com/Azure/azure-rest-api-specs-pr/pull/15270

**Comment by zaowang-ms** [comment]
Hi @mentat9, thanks for removing the tag for me. Do we need separate approvals for each SDK breaking change, or is the main one enough?

File: specification/nginx/resource-manager/NGINX.NGINXPLUS/stable/2023-09-01/swagger.json
Comment (mentat9) [inline specification/nginx/resource-manager/NGINX.NGINXPLUS/stable/2023-09-01/swagger.json:479]: >analysis [](http://example.com/codeflow?start=167&length=8)

This needs to be a verb, e.g. analyze.

File: specification/nginx/resource-manager/NGINX.NGINXPLUS/stable/2023-09-01/swagger.json
Comment (mentat9) [inline specification/nginx/resource-manager/NGINX.NGINXPLUS/stable/2023-09-01/swagger.json:1481]: >"x-ms-identifiers" [](http://example.com/codeflow?start=14&length=18)

You can identify the key order of the collection by adding [for example] virtualPath here.

File: specification/nginx/resource-manager/NGINX.NGINXPLUS/stable/2023-09-01/swagger.json
Comment (mentat9) [inline specification/nginx/resource-manager/NGINX.NGINXPLUS/stable/2023-09-01/swagger.json:1507]: >status [](http://example.com/codeflow?start=9&length=6)

Looks like this should be modeled as string enum (assuming it takes a fixed set of values).

File: specification/nginx/resource-manager/NGINX.NGINXPLUS/stable/2023-09-01/swagger.json
Comment (mentat9) [inline specification/nginx/resource-manager/NGINX.NGINXPLUS/stable/2023-09-01/swagger.json:1536]: >id [](http://example.com/codeflow?start=9&length=2)

Recommend against using property names the same as top-level properties in the standard ARM resource envelope. It's actually allowed in a POST payload, but I'd recommend errorId or similar instead of id.

**Comment by zaowang-ms** [comment]
Hello @mentat9,

Thank you for your thorough and insightful reviews, especially regarding "x-ms-identifiers" and "id."

We're currently at a stage where implementing changes is particularly challenging. Our workflow is structured such that modifications aren't anticipated during the synchronization from azure-rest-api-specs-pr to azure-rest-api-specs. This stage is critical for finalizing content that is already set to be released to the public.

Incorporating changes at this point would necessitate a return to the azure-rest-api-specs-pr stage, followed by initiating an entirely new PR. This process deviation could significantly delay our current release timeline. Also, introducing changes now could result in breaking changes for our customers, which is a critical concern, particularly regarding the verb analyze issue.

Given these constraints and the discussions about our workflow, we kindly request your approval to proceed with the current PR as is. We highly value your feedback and intend to include many of your suggestions in our next stable version. 

Also, below is a workflow of our API changes, based on our team's understanding of the api change workflow and accommodated with our business need to work with partners. Please let us know if this workflow is incorrect, especially the part where we do not anticipate changes during the synchronization from azure-rest-api-specs-pr to azure-rest-api-specs. Muchly appreciated!

image

**Comment by mentat9** [comment]
@zaowang-ms - If these changes were already reviewed and signed off by ARM in the private repo, just provide a link to the PR and I'll wave it through. We don't need to review your changes multiple times (especially if they are just moving from private to public repo).

Your flowchart above looks correct to me. In future, feel free to alert the ARM reviewer up front (including a link to the PR) if a change has already been reviewed and signed off by ARM in a different PR.

**Comment by zaowang-ms** [comment]
@mentat9 Thanks for replying! Yes, the changes were all approved and signedoff before when it was merged in the private repo. These are the two relevant PRs that included these changes:

https://github.com/Azure/azure-rest-api-specs-pr/pull/14498

https://github.com/Azure/azure-rest-api-specs-pr/pull/15270

I did mention it but should have provide these links earlier to reduce misunderstanding. Sorry for any delay and extra work caused by that, and thanks fir reviewing our workflow as well. Please let me know if anything else needed. Thanks!

**Comment by mentat9** [comment]
@zaowang-ms - Signing off for ARM per signoff in the private repo. For future APIs, please use verbs for POST actions.

**Comment by zaowang-ms** [comment]
Hi @mentat9 , thanks for the sign off. It's definitely something we should be careful of. Thanks again!

**Comment by zaowang-ms** [comment]
Notes for SDK reviewers:

The identified sdk breaking changes were all approved and signed off before when it was merged in the private repo. These are the two relevant PRs that included these changes:

https://github.com/Azure/azure-rest-api-specs-pr/pull/14498

https://github.com/Azure/azure-rest-api-specs-pr/pull/15270

As we've received ARMSignedOfff, I think we are awaiting SDK reviews now. Please let us know if any other information is needed from us. Thanks!

**Comment by zaowang-ms** [comment]
@rkmanda is there any extra approval needed for this PR? I've requested merge and there is no update yet for a day. It would be great if you can help to take a look on this, thanks!

**Comment by TimLovellSmith** [comment]
@mentat9 I still have some concerns about the verb thing. Waving it through means we'll never be able to fix it to a verb without it being a breaking change. And properly using verbs instead of nouns is actually quite useful to customers, for being able to understand the intent of the API - or also the meaning of the related RBAC action! 

I wouldn't say there is really anything wrong with the workflow, but just has a challenge here and now because of the reality that API review is neither perfect nor perfectly consistent.

Are we all okay with keeping it as a noun forever, in order to avoid breaking changes, just because of this workflow constraint? I.e. is it really only a convenience for ourselves? If so, I tend to think we should fix that one change in private-specs PR first and then sync it here.
I am also curious what some leaders think. /cc @JeffreyRichter @rkmanda

**Comment by mentat9** [comment]
> @mentat9 I still have some concerns about the verb thing. Waving it through means we'll never be able to fix it to a verb without it being a breaking change. And properly using verbs instead of nouns is actually quite useful to customers, for being able to understand the intent of the API - or also the meaning of the related RBAC action!
> 
> I wouldn't say there is really anything wrong with the workflow, but just has a challenge here and now because of the reality that API review is neither perfect nor perfectly consistent.
> 
> Are we all okay with keeping it as a noun forever, in order to avoid breaking changes, just because of this workflow constraint? And is the only reason not to fix it only a convenience for ourselves? If so, I tend to think we should fix that one change in private-specs PR first (smallest possible fix), and then sync it here. I am also curious what some leaders think. /cc @JeffreyRichter @rkmanda

I agree: that's the main thing that concerned me as well.

The same feedback was given on the original PR: https://github.com/Azure/azure-rest-api-specs-pr/pull/15270/files#r1366341718, but was never acted on (or responded to). Unless there was an agreement on some other channel, it seems like this one just slipped through the cracks.

I'm also interested in input from others (@JeffreyRichter, @mikekistler , @rkmanda).

**Comment by zaowang-ms** [comment]
Thanks for your input, @mentat9 and @TimLovellSmith. I'd like to share some thoughts regarding this:

1. The initial reason for using a noun in our API paths was to minimize differences with downstream APIs, specifically inheriting the path "analysis" provided by our partner. 
2. Why is it not responded at first place? If you examine the history of our private PR at https://github.com/Azure/azure-rest-api-specs-pr/pull/15270, you'll notice that the verb naming issue was raised after the PR was approved and merged. I think that at that time, the review queue was very long, and both the PR creators and reviewers were burdened, likely resulting in this oversight.
3. Despite this, the point about verb naming in the API is valid. There are only 2 main costs: 1. There will be a product breaking change as it’s already in use on Azure portal. 2. A delay of several weeks in introducing new features and their business impact.

Considering our discussion, we have two main options:
1. Revise the API to align with previous standards, collaborating closely to speed up the review and minimize associated costs.
2. Postpone this change, and plan for a breaking change in a future API version, reducing immediate business impact.

In short, we must balance the need for consistency with partner APIs and the efficiency of our review process. The choice between immediate revision or future planning will depend on how we prioritize these aspects in our API development strategy.

File: specification/nginx/resource-manager/NGINX.NGINXPLUS/stable/2023-09-01/swagger.json
Comment (JeffreyRichter) [reply specification/nginx/resource-manager/NGINX.NGINXPLUS/stable/2023-09-01/swagger.json:479]: It does seem like "analysis" should be "analyze" especially since the summary starts with "Analyze an NGINX configuration..." so the URL isn't consistent with its own documentation.
So, I agree this should be a noun.

If a version of the service has already shipped with "analysis" then changing it is a breaking change and we should not do it if any 3rd party customers could be using it. If the only customers are 1st party (like the Azure Portal), then we don't care about breaking 1st party as we can work with them closely to fix things without breaking customers and we always compromise 1st party to make a better long-term customer (3rd party) experience.

Also, we almost always compromise immediate negative business impact in favor of the long-term positive customer experience. If we could fix this now by introducing some delay (negative business impact) but not have to break customers in the future, then we should definitely do this. Breaking customers causes severe customer dissatisfaction with Azure all up - the breaking change board was created a few years ago within the Azure org with the sole charter of reducing all the breaks to improve overall Azure customer satisfaction.

**Comment by TimLovellSmith** [comment]
@zaowang-ms so from your side, you'd like to go ahead with the merge as-is?

**Comment by zaowang-ms** [comment]
> @zaowang-ms so from your side, you'd like to go ahead with the merge as-is?

Hi @TimLovellSmith , I think @JeffreyRichter commented as follows this morning and it didn't show up in the thread somehow:

> 
> It does seem like "analysis" should be "analyze" especially since the summary starts with "Analyze an NGINX configuration..." so the URL isn't consistent with its own documentation.
> So, I agree this should be a noun.
> 
> If a version of the service has already shipped with "analysis" then changing it is a breaking change and we should not do it if any 3rd party customers could be using it. If the only customers are 1st party (like the Azure Portal), then we don't care about breaking 1st party as we can work with them closely to fix things without breaking customers and we always compromise 1st party to make a better long-term customer (3rd party) experience.
> 
> Also, we almost always compromise immediate negative business impact in favor of the long-term positive customer experience. If we could fix this now by introducing some delay (negative business impact) but not have to break customers in the future, then we should definitely do this. Breaking customers causes severe customer dissatisfaction with Azure all up - the breaking change board was created a few years ago within the Azure org with the sole charter of reducing all the breaks to improve overall Azure customer satisfaction.


I appreciate Jeffrey's perspective, emphasizing long-term customer experience over immediate business concerns. We do initially prefer to proceed directly regarding the potential business cost as we work with F5 NGINX to expand the features of NGINXaaS for Azure quickly. However, now by understanding Jeffrey's points, we can adapt to this change and will communicate with our partners accordingly.

There are currently no 3rd party customers as the Api is still not public. I have created the PR to fix that: https://github.com/Azure/azure-rest-api-specs-pr/pull/16675 
This would be a breaking change and need approval again. Please allow me to ask for an expedited process on approve this one, if possible, to reduce impact on feature delivery schedules and 1st services. Thanks!

**Comment by TimLovellSmith** [comment]
I can remove mergeRequested for now until we fixed the other PR then, I think, then we can add it back.

**Comment by zaowang-ms** [comment]
Now it's a good time to merge it as it's been changed to /analyze .

**Review by raosuhas — APPROVED** [review]

**Pull request merged**
